### PR TITLE
Merge Task Steps Service Layer and

### PR DESCRIPTION
**Pull request by agaldanaw2**
GitHub Issue (If applicable): #109  #131 

## PR Type

What kind of change does this PR introduce?
- Feature

## What is the current behavior?
The task list display name cannot be updated and We can't add steps to taks

## What is the new behavior?

Support management of task steps on service layer and add functionality for renaming of task list

## PR Checklist

Please check if your PR fulfills the following requirements:

- [ ] Tested code with current [supported SDKs](../README.md#supported)
- [ ] Docs have been added/updated which fit [documentation template](https://github.com/nventive/Uno/blob/master/doc/.feature-template.md). (for bug fixes / features)
- [ ] [Unit Tests and/or UI Tests](doc/articles/working-with-the-samples-apps.md) for the changes have been added (for bug fixes / features) (if applicable)
- [ ] [Wasm UI Tests](doc/articles/working-with-the-samples-apps.md#running-the-webassembly-ui-tests-snapshots) are not showing unexpected any differences. Validate PR `Screenshots Compare Test Run` results.
- [ ] Contains **NO** breaking changes
- [ ] Updated the [Release Notes](https://github.com/nventive/Uno/tree/master/doc/ReleaseNotes)
- [ ] Associated with an issue (GitHub or internal)

### CODE REVIEW

File: src/ToDo.UI/ToDo.UI.projitems [hunk 1]
```diff
@@ -158,6 +155,5 @@
     <_Globbed_Content Include="$(MSBuildThisFileDirectory)Assets/**/*.*" Exclude="@(Content)" />
     <Content Include="@(_Globbed_Content)" />
     <_Globbed_Embedded_Resource Include="$(MSBuildThisFileDirectory)*.json" Exclude="@(EmbeddedResource)" />
-    <EmbeddedResource Include="@(_Globbed_Embedded_Resource)" />
```
Comment (nickrandolph) [inline src/ToDo.UI/ToDo.UI.projitems:161]: Make sure this line is added back in

File: src/ToDo.UI/ToDo.UI.shproj [hunk 2]
```diff
@@ -10,5 +10,8 @@
   <PropertyGroup />
   <Import Project="ToDo.UI.projitems" Label="Shared" />
   <Import Project="$(MSBuildExtensionsPath32)\Microsoft\VisualStudio\v$(VisualStudioVersion)\CodeSharing\Microsoft.CodeSharing.CSharp.targets" />
-
+  <ItemGroup>
```
Comment (nickrandolph) [inline src/ToDo.UI/ToDo.UI.shproj:13]: Why were these removed?

File: src/ToDo.UI/ToDo.UI.projitems (hunk 1 above)
Comment (agaldanaw2) [reply src/ToDo.UI/ToDo.UI.projitems:161]: sure

File: src/ToDo.UI/ToDo.UI.shproj [hunk 3]
```diff
@@ -11,4 +11,4 @@
   <Import Project="ToDo.UI.projitems" Label="Shared" />
   <Import Project="$(MSBuildExtensionsPath32)\Microsoft\VisualStudio\v$(VisualStudioVersion)\CodeSharing\Microsoft.CodeSharing.CSharp.targets" />
 
-</Project>
\ No newline at end of file
+</Project>
```
Comment (agneszitte-nventive) [inline src/ToDo.UI/ToDo.UI.shproj:14]: Why this change? Can be removed from your PR no?

File: src/ToDo.UI/Views/Dialogs/RenameListFlyout.xaml [hunk 4]
```diff
@@ -1,50 +1,54 @@
...
-			<Button Content="Rename"
-					uen:Navigation.Request="-"
-					uen:Navigation.Data="{Binding Text, ElementName=ListNameText}"
-					utu:AutoLayout.CounterAlignment="Start"
-					Style="{StaticResource MaterialContainedButtonStyle}" />
-
-		</utu:AutoLayout>
-
-	</utu:AutoLayout>
+﻿<Page
```
Comment (agneszitte-nventive) [inline src/ToDo.UI/Views/Dialogs/RenameListFlyout.xaml:1]: What is the real change here? Please update the xaml for only the lines you made changes and not the entire page.

File: src/ToDo.UI/Views/Dialogs/RenameListFlyout.xaml [hunk 5]
```diff
@@ -1,50 +1,54 @@
...
+                Spacing="10" />
+
+            <Button
+                uen:Navigation.Request="-"
+                Content="Cancel"
+                Style="{StaticResource MaterialTextButtonStyle}" />
+
+            <Button
+                utu:AutoLayout.CounterAlignment="Start"
+                Command="{Binding RenameCommand}"
```
Comment (agneszitte-nventive) [inline src/ToDo.UI/Views/Dialogs/RenameListFlyout.xaml:47]: There is no need for the RenameCommand anymore normally as it's done by the extension with the navigation normally
```
uen:Navigation.Request="-"
uen:Navigation.Data="{Binding Text, ElementName=ListNameText}"
```

File: src/ToDo/Presentation/Dialogs/RenameListViewModel.cs [hunk 6]
```diff
@@ -2,5 +2,21 @@ namespace ToDo.Presentation;
 
 public partial class RenameListViewModel
 {
-    
-}
\ No newline at end of file
+	private INavigator Navigator { get; }
```
Comment (agneszitte-nventive) [inline src/ToDo/Presentation/Dialogs/RenameListViewModel.cs:5]: There is no need for the RenameCommand anymore normally as it's done by the extension with the navigation normally in [RenameListFlyout.xaml](https://github.com/unoplatform/uno.todo/pull/150/files#diff-655db2e39acf518fbaa3b870cde099dd94a39d131ff35fba91e1195afaab1d3f)

```
uen:Navigation.Request="-"
uen:Navigation.Data="{Binding Text, ElementName=ListNameText}"
````

File: src/ToDo.UI/lists.json [hunk 7]
```diff
@@ -23,4 +23,4 @@
     "wellknownListName": "none",
     "id": "List-Groceries"
   }
-]
+]
```
Comment (agneszitte-nventive) [inline src/ToDo.UI/lists.json:26]: Why this change? Can be removed from your PR no?

File: src/ToDo/Business/TaskService.cs [hunk 8]
```diff
@@ -69,4 +69,44 @@ public async ValueTask<IImmutableList<ToDoTask>> GetAllAsync(string displayName
...
 	}
+
+	/// <inheritdoc />
+
+	public async Task AddStepAsync(string baseTaskListId, string baseTaskId,
+		CheckListItem checkListItem, CancellationToken ct)
+	{
+		var createdStep = await _client.AddStepAsync(baseTaskListId, baseTaskId, checkListItem.ToData(), ct);
+
+		//TODO: add messenger notification
```
Comment (agneszitte-nventive) [inline src/ToDo/Business/TaskService.cs:80]: Do you have an issue/item in the backlog for the 3 To-Do(s) you added in TaskService  ?

File: src/ToDo/Presentation/TaskListViewModel.cs [hunk 9]
```diff
@@ -113,12 +113,11 @@ private async ValueTask RenameList(TaskList list, CancellationToken ct)
 		}
 
 		var newListName = (await response.Result).SomeOrDefault();
-		if (!String.IsNullOrWhiteSpace(newListName))
+		if ( !String.IsNullOrWhiteSpace(newListName))
```
Comment (agneszitte-nventive) [inline src/ToDo/Presentation/TaskListViewModel.cs:116]: This change is not needed here no ?

**Review by agneszitte-nventive — CHANGES_REQUESTED** [review]

File: src/ToDo.UI/Views/Dialogs/RenameListFlyout.xaml (hunk 4 above)
Comment (Xiaoy312) [reply src/ToDo.UI/Views/Dialogs/RenameListFlyout.xaml:1]: Also, please grab the XamlStyler visual-studio extension for formatter.

**Review by Xiaoy312 — COMMENTED** [review]
there is a lot of commit with "save" as title
please clean them up the commits with `rebase -i`

File: src/ToDo/Business/ITaskService.cs [hunk 10]
```diff
@@ -13,4 +13,13 @@ public interface ITaskService
...
 
 	ValueTask<IImmutableList<ToDoTask>> GetAllAsync(string displayName = "", CancellationToken ct = default);
+
+	Task AddStepAsync(string baseTaskListId, string baseTaskId, CheckListItem checkListItem, CancellationToken ct);
+
+	ValueTask<IImmutableList<CheckListItem>> GetStepsAsync(string baseTaskListId, string baseTaskId, CancellationToken ct);
+
+	Task DeleteStepAsync(string baseTaskListId, string baseTaskId, string checklistItemId, CancellationToken ct);
+
+	Task UpdateStepAsync(string baseTaskListId, string baseTaskId, CheckListItem checkListItem, CancellationToken ct);
```
Comment (kazo0) [inline src/ToDo/Business/ITaskService.cs:23]: nit: I feel like these shouldn't have "Async" as a suffix, unless there are non-async versions of these already. Not sure what the accepted convention is for the unoplatform org

File: src/ToDo.UI/Views/Dialogs/RenameListFlyout.xaml (hunk 4 above)
Comment (agaldanaw2) [reply src/ToDo.UI/Views/Dialogs/RenameListFlyout.xaml:1]: I didn't change entire, using XamlStyler and save the changes was made it. The real change was add the command for renaming of the list

File: src/ToDo.UI/Views/Dialogs/RenameListFlyout.xaml (hunk 5 above)
Comment (agaldanaw2) [reply src/ToDo.UI/Views/Dialogs/RenameListFlyout.xaml:47]: Okay, thanks

File: src/ToDo/Business/ITaskService.cs (hunk 10 above)
Comment (agaldanaw2) [reply src/ToDo/Business/ITaskService.cs:23]: It's convention for unoplatform.

File: src/ToDo/Business/TaskService.cs (hunk 8 above)
Comment (agaldanaw2) [reply src/ToDo/Business/TaskService.cs:80]: It's only a Task, service layer for task step

File: src/ToDo.UI/lists.json (hunk 7 above)
Comment (agaldanaw2) [reply src/ToDo.UI/lists.json:26]: yeap

File: src/ToDo.UI/ToDo.UI.shproj (hunk 3 above)
Comment (agaldanaw2) [reply src/ToDo.UI/ToDo.UI.shproj:13]: I don't know, I forgot discard this changes

File: src/ToDo/Presentation/TaskListViewModel.cs (hunk 9 above)
Comment (agaldanaw2) [reply src/ToDo/Presentation/TaskListViewModel.cs]: No, it was discarted

File: src/ToDo.UI/ToDo.UI.projitems [hunk 11]
```diff
@@ -14,10 +14,6 @@
     </Content>
     <None Include="$(MSBuildThisFileDirectory)Strings\Resources.l12n.xml" />
   </ItemGroup>
-  <ItemGroup>
```
Comment (nickrandolph) [inline src/ToDo.UI/ToDo.UI.projitems:17]: Should these be removed?

File: src/ToDo/Business/CheckListItem.cs [hunk 12]
```diff
@@ -0,0 +1,40 @@
+﻿namespace ToDo;
+
+public class CheckListItem
```
Comment (nickrandolph) [inline src/ToDo/Business/CheckListItem.cs:3]: Can this be a record?

File: src/ToDo/Business/ITaskService.cs [hunk 13]
```diff
@@ -13,4 +13,13 @@ public interface ITaskService
 	ValueTask<IImmutableList<ToDoTask>> GetAsync(TaskList list, CancellationToken ct);
 
 	ValueTask<IImmutableList<ToDoTask>> GetAllAsync(string displayName = "", CancellationToken ct = default);
+
+	Task AddStepAsync(string baseTaskListId, string baseTaskId, CheckListItem checkListItem, CancellationToken ct);
```
Comment (nickrandolph) [inline src/ToDo/Business/ITaskService.cs:17]: Can we return ValueTask for these methods

File: src/ToDo/Data/CheckListItemData.cs [hunk 14]
```diff
@@ -0,0 +1,22 @@
+﻿namespace ToDo;
+
+public class CheckListItemData
```
Comment (nickrandolph) [inline src/ToDo/Data/CheckListItemData.cs:3]: Can this be a record?

File: src/ToDo/Data/Mock/MockTaskListEndpoint.cs [hunk 15]
```diff
@@ -142,5 +144,54 @@ public async Task<TaskReponseData<TaskData>> GetAllTasksAsync(string displayName
...
+		var list = await LoadListTasks(baseTaskListId);
+		var task = list.FirstOrDefault(t => t.Id == baseTaskId);
+		if (task is not null)
+		{
+			if(task.ChecklistItems is null)
+			{
+				task.ChecklistItems = new List<CheckListItemData>();
+			}
 
+			task.ChecklistItems?.Add(checkListItem);
```
Comment (nickrandolph) [inline src/ToDo/Data/Mock/MockTaskListEndpoint.cs:159]: Remove nullable check - ChecklistItems can't be null here

File: src/ToDo/Data/Mock/MockTaskListEndpoint.cs [hunk 16]
```diff
@@ -142,5 +144,54 @@ public async Task<TaskReponseData<TaskData>> GetAllTasksAsync(string displayName
...
+			return checkListItem;
+		}
+		return new CheckListItemData();
+	}
+
+	public async Task<TaskReponseData<CheckListItemData>> GetStepsAsync(string baseTaskListId, string baseTaskId, CancellationToken ct)
+	{
+		var list = await LoadListTasks(baseTaskListId);
+		var task = list.FirstOrDefault(t => t.Id == baseTaskId);
+		if (task is not null)
```
Comment (nickrandolph) [inline src/ToDo/Data/Mock/MockTaskListEndpoint.cs:169]: combine with previous line eg
if (list.FirstOrDefault(t => t.Id == baseTaskId) is { } task)

File: src/ToDo/Data/Mock/MockTaskListEndpoint.cs [hunk 17]
```diff
@@ -142,5 +144,54 @@ public async Task<TaskReponseData<TaskData>> GetAllTasksAsync(string displayName
...
+		{
+			return new TaskReponseData<CheckListItemData> { Value = task.ChecklistItems };
+		}
+		
+		return new TaskReponseData<CheckListItemData> { Value = Enumerable.Empty<CheckListItemData>().ToList() };
+	}
+	public async Task DeleteStepAsync(string baseTaskListId, string baseTaskId, string checklistItemId, CancellationToken ct)
+	{
+		var list = await LoadListTasks(baseTaskListId);
+		var task = list.FirstOrDefault(t => t.Id == baseTaskId);
```
Comment (nickrandolph) [inline src/ToDo/Data/Mock/MockTaskListEndpoint.cs:179]: See previous comment about combining these lines

File: src/ToDo/Data/Mock/MockTaskListEndpoint.cs [hunk 18]
```diff
@@ -142,5 +144,54 @@ public async Task<TaskReponseData<TaskData>> GetAllTasksAsync(string displayName
...
+		if (task is not null)
+		{
+			var step = task.ChecklistItems?.FirstOrDefault(x => x.Id == checklistItemId);
+			if(step is not null)
+			{
+				task.ChecklistItems?.Remove(step);
```
Comment (nickrandolph) [inline src/ToDo/Data/Mock/MockTaskListEndpoint.cs:185]: null check not required - ChecklistItems can't be null at this point

**Review by nickrandolph — APPROVED** [review]
Actually, the lines that remove appsettings.json need to be reverted

**Review by nickrandolph — APPROVED** [review]

File: src/ToDo/Business/CheckListItem.cs (hunk 12 above)
Comment (agaldanaw2) [reply src/ToDo/Business/CheckListItem.cs:3]: Sure

File: src/ToDo/Business/ITaskService.cs (hunk 13 above)
Comment (agaldanaw2) [reply src/ToDo/Business/ITaskService.cs:17]: Sure

File: src/ToDo/Data/CheckListItemData.cs (hunk 14 above)
Comment (agaldanaw2) [reply src/ToDo/Data/CheckListItemData.cs:3]: sure

**Comment by nickrandolph** [comment]
I've rebased this branch to main, and squashed all commits to make this PR easier to follow. I've also marked this PR as draft because it's not ready to be merged
@francoistanguay / @carldebilly can we decide whether we want to add support for task steps? If so, we'll need to add the UI etc. If we're not going to add steps, we can close this PR

**Comment by nickrandolph** [comment]
I'm going to close this PR - if we want to add steps down the line we can reopen this PR
